### PR TITLE
pr

### PR DESCRIPTION
**Pull request by ThinhNgVhust**
sometime facing hard enough problems. maybe it is easy for geeks or someone else. It is the time i doubting myself. think that i am not good enough. But just keep goin....

### CODE REVIEW

**Comment by trongtai37** [comment]
Hi @ThinhNgVhust, I do not know much a bout Python but I have some suggestions and hoping they will make your code cleaner and easier to read.
- You can consider using a code formatter for Python, if you are using VSCode, you can refer `ms-python.python` extension.
- You should follow some conventions of writing code in python like snake_case for variables, functions; CamelCase for classes, ...
Hoping it helps you :D

**Comment by ThinhNgVhust** [comment]
Hi, @trongtai37 First of all, thank you for your comment. Can you show me where I was not following the coding convention? Sometime i use "camelCase", but sometimes "camel_case" instead. 
Please feel free to talk about this

File: homework/thinhbka/lesson-4-hashing/1_Subarray_Sums_Divisible_by_K.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
+#Create prefixS array :prefixS[i] = arr[0] + arr[1]+ ... + arr[i]
+# sum(i,j) = arr[i] + arr[i+1] + ... + arr[j] = prefixS[j]- prefixS[i-1]
+# sum(i,j) mod k == 0  => (pS[j]-ps[i]) mod k == 0 => pS[j] mod k == pS[i-1] mod k.
+
+class Solution:
+    def subarraysDivByK(self, nums: List[int], k: int) -> int:
```
Comment (trongtai37) [inline homework/thinhbka/lesson-4-hashing/1_Subarray_Sums_Divisible_by_K.py:7]: Hi @ThinhNgVhust , for example the function name `subarraysDivByK` should be `sub_arrays_div_by_k`

Comment (ThinhNgVhust) [reply homework/thinhbka/lesson-4-hashing/1_Subarray_Sums_Divisible_by_K.py:7]: hi @trongtai37 , really appreciate for your comment. But it is seemly like default Leetcode's coding convention. 
You could try to switch a problem on Lc to python :((

Comment (trongtai37) [reply homework/thinhbka/lesson-4-hashing/1_Subarray_Sums_Divisible_by_K.py:7]: Yes, let's skip this kind of issues :D

File: homework/thinhbka/lesson-4-hashing/2_Maximum_Subarray.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# arr  = [-1, 4, -1, 7, 8]
+
+#sum(i,j) = preS[j] - preS[i-1] 
+# sum(i,j) max <=> preS[j] is biggest, and preS[i-1] smallest, i-1 <= j
+class Solution:
+    def maxSubArray(self, nums: List[int]) -> int:
+        preS =[0]
```
Comment (trongtai37) [inline homework/thinhbka/lesson-4-hashing/2_Maximum_Subarray.py:7]: How about this one @ThinhNgVhust, I think the variable's name should be `pre_sum`

File: homework/thinhbka/lesson-4-hashing/2_Maximum_Subarray.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        for e in nums:
+            preS.append(e+preS[-1])
+        ans = -1e9
+        minSoFar = 0
```
Comment (trongtai37) [inline homework/thinhbka/lesson-4-hashing/2_Maximum_Subarray.py:11]: And so is this variable, `min_so_far` or `min_pre_sum_from_begining`

File: homework/thinhbka/lesson-4-hashing/3_4Sum.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    def fourSum(self, nums, target):
+        n= len(nums)
+        nums.sort()
+        # print("hash table ",sum12)
+        ans = set()
+        for a in range(n):
+            for b in range(a+1,n):
+                s = nums[a] + nums[b]
+                L = b+1
+                R = n-1
```
Comment (trongtai37) [inline homework/thinhbka/lesson-4-hashing/3_4Sum.py:49]: and these variables should be written in lowercase
```suggestion
                l = b+1
                r = n-1
```

Comment (ThinhNgVhust) [reply homework/thinhbka/lesson-4-hashing/3_4Sum.py:49]: oh, It is my bad :(

File: homework/thinhbka/lesson-4-hashing/4_Insert_Delete_GetRandom_O(1).py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            del self.dic[val]
+            return True
+        return False
+
+    def getRandom(self) -> int:
+        """
+        Get a random element from the set.
+        """
+     
+        return random.choice(list(self.dic.values()))
```
Comment (quanghoang33) [inline homework/thinhbka/lesson-4-hashing/4_Insert_Delete_GetRandom_O(1).py]: will it cost O(n)?

Comment (ThinhNgVhust) [reply homework/thinhbka/lesson-4-hashing/4_Insert_Delete_GetRandom_O(1).py]: Hi, @quanghoang33. Although this solution still AC in LC, but it is definitely take O(n) for get **random**. 
I will change this code and PR latter.
Regards

Comment (ThinhNgVhust) [reply homework/thinhbka/lesson-4-hashing/4_Insert_Delete_GetRandom_O(1).py]: Changed

**Pull request merged**